Button code clean up and bevel fixes

sam-b-rose opened the pull request:
Closes [#11191](https://github.com/Shopify/polaris/issues/11191)

Closes #11162

This PR:

- Is a rewrite of all `Button.scss` styles reducing the SCSS from `>1k` lines to `287`
- This also removes a few unnecessary internal DOM elements. Why should button have so many elements 🤷 
- Fixes the bevels for all button variants 👈  the main point of the original issue (thanks @heyjoethomas for the design help!)
- Also fixes the Secondary Actions buttons in the Page component

This greatly reduces the complexity of the Button styles and focuses on readability of the custom property variables.

🎩 with this 🌀[spin url](https://admin.web.button.sam-rose.us.spin.dev/store/shop1)

![localhost_6006_iframe html_id=all-components-button--all viewMode=story(iPad Mini)](https://github.com/Shopify/polaris/assets/11774595/8b592fa9-dfb2-4dc1-b451-950ca130c389)

sam-b-rose committed:
Button code clean up and bevel refactoring

sam-b-rose committed:
Update Page action button overrides

alex-page commented on polaris-react/src/components/Button/Button.tsx:
```diff
@@ -114,48 +112,46 @@ export function Button({
   dataPrimaryLink,
   tone,
-  variant,
+  variant = 'secondary',
```
Could this be a breaking change if a `<Button>` without variant today potentially changes it's appearance? I can't quickly see where the default value for `variant` is or was..

sam-b-rose replied:
I don't _think_ so. If a Button didn't have a variant prop, it would internally map `undefined` to now be `secondary`. This will then apply the "default" styles that where being applied for `<Button variant={undefined}>` before.

If we don't do it this way, then we have to apply all the secondary styles as the default styles. The reason I didn't do this is so we could avoid having to override every style used to create the `secondary` variant in other variants that don't need certain styles (e.g., `plain` doesn't need a background color)

chloerice replied:
I believe `secondary` is being added as a variant to make the default button styles an explicit `variant`, so it will be the same as a Button with no `variant` set.

chloerice commented on .changeset/tender-pugs-crash.md:
```diff
@@ -0,0 +1,5 @@
+'@shopify/polaris': minor
+---
+
+Refactor Button styles and bevel
```
```suggestion
Refactored `Button` styles and bevel
```

chloerice commented on polaris-react/src/components/Button/Button.scss:
```diff
@@ -1,1043 +1,286 @@
-  --pc-button-color-hover: var(--p-color-bg-fill-hover);
-  --pc-button-color-active: var(--p-color-bg-fill-active);
-  --pc-button-color-depressed: var(--p-color-bg-fill-active);
+.toneSuccess:is(.variantPrimary) {
```
Unfortunately I don't think we are able to use [`:is`](https://caniuse.com/?search=%3Ais) with our [current minimum browser support](https://github.com/Shopify/web/blob/main/.browserslistrc) 👀

alex-page replied:
We currently use `:is` ~20 times in Polaris 💔 . I also opened this PR in shopify/web https://github.com/Shopify/web/pull/111914

chloerice replied:
<img src="https://media.giphy.com/media/oJGqcAvg5stm8/giphy.gif" alt="Spongebob Squarepants dancing in celebration with a marching band" />

aaronccasanova replied:
Heads up, this is the result on Safari 14 in BrowserStack (wasn't able to load Safari 13). cc @sam-b-rose @chloerice @alex-page

**Before**
<img width="695" alt="Screenshot 2023-12-15 at 8 38 39 AM" src="https://github.com/Shopify/polaris/assets/32409546/f843718a-6dd9-4fcb-bfd8-3e611cba846d">


**After**
<img width="630" alt="Screenshot 2023-12-15 at 8 29 17 AM" src="https://github.com/Shopify/polaris/assets/32409546/68ec8e52-e0eb-4ff1-ac6d-b056e56a0e20">

chloerice commented on polaris-react/src/components/Button/Button.scss:
```diff
@@ -1,1043 +1,286 @@
-    margin: calc(-1 * var(--p-space-100));
-
-    path {
-      fill: var(--p-color-icon-secondary);
```
The `iconOnly` `fill` declarations will need to be ported over to the rewritten styles, with this missing they're falling back to `currentColor`:

| Current | Rewrite |
|--------|--------|
| <img width="1512" alt="Screenshot 2023-12-07 at 3 27 40 PM" src="https://github.com/Shopify/polaris/assets/18447883/25c7af81-20a2-4d59-ac3e-66692194e4f6"> | <img width="1512" alt="Screenshot 2023-12-07 at 3 24 12 PM" src="https://github.com/Shopify/polaris/assets/18447883/0cff4e01-29c5-413c-a741-2dae30b18e7f"> |
|<img width="1512" alt="Screenshot 2023-12-07 at 3 59 25 PM" src="https://github.com/Shopify/polaris/assets/18447883/a43f05c4-a863-498a-9623-b3aba0cfda87">| <img width="1512" alt="Screenshot 2023-12-07 at 3 32 02 PM" src="https://github.com/Shopify/polaris/assets/18447883/c0f31383-fd74-4643-ad5f-021cf56255b9">|
|<img width="1512" alt="Screenshot 2023-12-07 at 3 55 57 PM" src="https://github.com/Shopify/polaris/assets/18447883/25e38cb8-bc0f-4fad-9d6e-2576471f7b67">|<img width="1512" alt="Screenshot 2023-12-07 at 3 57 03 PM" src="https://github.com/Shopify/polaris/assets/18447883/b631e085-8b6b-4c3d-a06e-d1cbb8dc6e8f">|
|<img width="1512" alt="Screenshot 2023-12-07 at 4 02 25 PM" src="https://github.com/Shopify/polaris/assets/18447883/6627b0e7-8c14-4c18-9629-ff642a1e8a03">|<img width="1512" alt="Screenshot 2023-12-07 at 4 02 12 PM" src="https://github.com/Shopify/polaris/assets/18447883/d8721693-5ea0-42fd-9a40-f0d61f01ac7b">|

chloerice commented on polaris-react/src/components/Button/Button.scss:
```diff
@@ -1,1043 +1,286 @@
+  --color_hover: var(--color);
+  --color_active: var(--color);
+  --color_pressed: var(--color_active);
+  --color_disabled: var(--p-color-text-disabled);
```
We'll need to update the overrides of the Button `--pc-button-` properties in [`ContextualSaveBar`](https://github.com/Shopify/polaris/blob/2b5ecf8df3dcc07a3a6a75939e6f3b588f710ea1/polaris-react/src/components/Frame/components/ContextualSaveBar/ContextualSaveBar.scss#L84-L93)

| Current | Rewrite |
|--------|--------|
| <img width="1512" alt="Screenshot 2023-12-07 at 3 36 22 PM" src="https://github.com/Shopify/polaris/assets/18447883/f44123af-8e42-4853-bf67-0f991a261141"> | <img width="1512" alt="Screenshot 2023-12-07 at 3 38 10 PM" src="https://github.com/Shopify/polaris/assets/18447883/2649011e-3dee-4ea3-a487-492bcffc9b96">|

chloerice reviewed (dismissed):
The `Button` rewrite looks and works amazing @sam-b-rose, the SCSS is so easy to read and reason about now!! 🔥  The only bugs I found while tophatting were in icon only plain/tertiary/monochromePlain buttons and needing to update the Button overrides in `ContextualSaveBar`. Updated the PR description with a working Spinstance.

sam-b-rose committed:
Remove linear gradient on secondary, success, and critical

sam-b-rose committed:
Update changeset message

sam-b-rose committed:
Fix icon colors and add --pc-* prefix

sam-b-rose committed:
Update ContextualSaveBar button overrides

sam-b-rose committed:
Improve `pressed` state for tertiary Button

The `pressed` state for tertiary buttons is currently unsatisfactory. Only a faint box-shadow appears when pressed. It should use the `--p-color-bg-fill-selected` color instead.

Co-Authored-By: Marc Thomas <hello@mrcthms.com>

sam-b-rose committed:
Allow custom component properties (fix lint)

sam-b-rose committed:
Adjust button height to be larger on mobile

chloerice commented on polaris-react/src/components/Button/Button.scss:
```diff
@@ -1,1043 +1,317 @@
+  background: var(--pc-button-bg);
+  color: var(--pc-button-color);
+  box-shadow: var(--pc-button-box-shadow);
+  fill: var(--pc-button-icon-fill);
```
```suggestion
```

chloerice commented on polaris-react/src/components/Button/Button.scss:
```diff
@@ -1,1043 +1,317 @@
-  transition: transform 75ms cubic-bezier(0.19, 0.91, 0.38, 1);
 
-  @media #{$p-breakpoints-md-up} {
+  @media (--p-breakpoints-md-up) {
```
```suggestion
  
  svg {
    fill: var(--pc-button-icon-fill);
  }

  @media (--p-breakpoints-md-up) {
```

chloerice commented on polaris-react/src/components/Button/Button.scss:
```diff
@@ -1,1043 +1,317 @@
+.Button:disabled,
+.Button[disabled],
+.disabled {
+  background: var(--pc-button-bg_disabled);
```
```suggestion
  --pc-button-icon-fill: var(--pc-button-icon-fill_disabled);
  background: var(--pc-button-bg_disabled);
```

chloerice commented on polaris-react/src/components/Button/Button.scss:
```diff
@@ -1,1043 +1,317 @@
+.disabled {
+  background: var(--pc-button-bg_disabled);
+  color: var(--pc-button-color_disabled);
+  fill: var(--pc-button-icon-fill_disabled);
```
```suggestion
```

chloerice commented on polaris-react/src/components/Button/Button.scss:
```diff
@@ -1,1043 +1,317 @@
-  position: relative;
+.sizeLarge {
+  --pc-button-padding-block: var(--p-space-150);
+  --pc-button-padding-inline: var(--p-space-300);;
```
```suggestion
  --pc-button-padding-inline: var(--p-space-300);
```

chloerice commented on polaris-react/src/components/Button/Button.scss:
```diff
@@ -1,1043 +1,317 @@
-    clip-path: none !important;
-  }
+// INTERACTION
+.pressable:active:not(.variantTertiary, .variantPlain, .variantMonochromePlain)
```
```suggestion
/* stylelint-disable-next-line selector-max-specificity -- Unchained :not selectors don't compile as intended, see https://github.com/Shopify/polaris/pull/9556) */
.pressable:active:not(.variantTertiary):not(.variantPlain):not(.variantMonochromePlain)
```

sam-b-rose replied:
I'm `:not` sure if this is a bug that was fixed, but checking the CSS build output, it appears to be compiling correctly 👍 

```css
/* Looking at polaris/polaris-react/build/esm/styles.css */

.Polaris-Button--pressable:active:not(.Polaris-Button--variantTertiary, .Polaris-Button--variantPlain, .Polaris-Button--variantMonochromePlain) > * { transform: translate3d(0, 0.0625rem, 0); }
```

chloerice commented on polaris-react/src/components/Button/Button.scss:
```diff
@@ -1,1043 +1,317 @@
-  .Icon:only-child {
-    margin-right: 0;
-  }
+  --pc-button-padding-inline: var(--p-space-100);
```
The tertiary and plain icon-only buttons are still black and blue:

| Before ([Chromatic](https://5d559397bae39100201eedc1-tfqrgrlkfi.chromatic.com/?path=/story/all-components-button--all)) ([Spin](https://admin.web.test-button-rewrite.chloe-rice.us.spin.dev/store/shop1/products/8))| After ([Chromatic](https://5d559397bae39100201eedc1-vtvnmjbofa.chromatic.com/?path=/story/all-components-button--all)) ([Spin](https://admin.web.test-revamp-button-fixes.chloe-rice.us.spin.dev/store/shop1/products/6))|
|--------|--------|
| ![Screenshot 2023-12-11 at 11 04 20 PM](https://github.com/Shopify/polaris/assets/18447883/cd6370f3-f619-4786-a74a-94af3cf190df) |![Screenshot 2023-12-11 at 11 09 56 PM](https://github.com/Shopify/polaris/assets/18447883/3c53b147-ed75-4e53-a0ae-b46a2c84afc8)|
| ![Screenshot 2023-12-11 at 11 04 42 PM](https://github.com/Shopify/polaris/assets/18447883/1bf6cd20-e532-4cef-8a8c-60b3df483d91)|![Screenshot 2023-12-12 at 12 35 43 AM](https://github.com/Shopify/polaris/assets/18447883/0eb36386-4c64-4c24-a6b3-fa3200e847e8) |
| ![Screenshot 2023-12-11 at 11 04 03 PM](https://github.com/Shopify/polaris/assets/18447883/743b6fb6-031c-4ca9-a745-3386e17493de) |![Screenshot 2023-12-11 at 11 09 44 PM](https://github.com/Shopify/polaris/assets/18447883/857f6073-7e49-4f3e-992d-3fe7a38a5b86)|

```suggestion
  &.variantTertiary,
  &.variantPlain {
    --pc-button-icon-fill: var(--p-color-icon-secondary);

    /* stylelint-disable-next-line -- selector-max-class */
    &.toneCritical {
      --pc-button-icon-fill: var(--p-color-text-critical);
    }
  }
}

/* stylelint-disable selector-max-specificity -- Need the high specificity for icon-only interation states because of how high specificity is with :is usage in the variant and tone class declarations above */
.iconOnly:is(:hover, :focus):not(.variantPrimary):not(.toneCritical.variantTertiary) {
  --pc-button-icon-fill: var(--p-color-icon-hover);
}

.iconOnly:is(:active, .pressed):not(.variantPrimary):not(.toneCritical.variantTertiary) {
  --pc-button-icon-fill: var(--p-color-icon-active);
}
/* stylelint-enable selector-max-specificity */

.disabled:not(.variantPrimary),
.iconOnly.disabled:not(.variantPrimary) {
  --pc-button-icon-fill_disabled: var(--p-color-icon-disabled);
  --pc-button-icon-fill: var(--pc-button-icon-fill_disabled);
}

.iconOnly:is(.sizeSlim, .sizeMedium) {
  --pc-button-padding-inline: var(--p-space-100);
```

sam-b-rose replied:
Good catch, updated 👍

chloerice replied:
Plain and tertiary are fixed! 🙌🏽 Just still need to override/handle the toneCritical plain and tertiary variants:

Tertiary critical icon only should be filled with `color-text-critical-x` tokens
![Screenshot 2023-12-14 at 11 47 41 AM](https://github.com/Shopify/polaris/assets/18447883/f3b96fa5-7a76-4d69-a2ff-1e001b516d19)

Plain critical icon only has the right interaction states but needs `color-text-critical` when not being interacted with
![Screenshot 2023-12-14 at 11 47 50 AM](https://github.com/Shopify/polaris/assets/18447883/0b5148b1-14e8-49c7-adec-c910742629e0)

chloerice reviewed (changes requested):
ContextualSaveBar looks great! Those pesky icon-only buttons still need the correct fill and interaction states. I suggested tophatted fixes and did my best to finesse the CSS needed to fit the new conventions 😅

mrcthms reviewed (approved):
LGTM for the tertiary pressed issue

sam-b-rose committed:
Apply fill to child SVG element

sam-b-rose committed:
Use icon secondary color for tertiary buttons

chloerice commented on polaris-react/src/components/Button/Button.scss:
```diff
@@ -1,1043 +1,352 @@
-  }
+.iconOnly:is(.variantTertiary) {
+  --pc-button-icon-fill: var(--p-color-icon-secondary);
+  --pc-button-icon-fill_hover: var(--p-color-icon);
```
Should the pressed/active and hover states of tertiary icon only buttons use the `icon-secondary-x` tokens?

```suggestion
  --pc-button-icon-fill_hover: var(--p-color-icon-secondary-hover);
  --pc-button-icon-fill_active: var(--p-color-icon-secondary-active);
```

|| `--p-color-icon-x` | `--p-color-icon-secondary-x` |
|--------|--------|--------|
|(static for comparison)|![Screenshot 2023-12-13 at 12 30 12 PM](https://github.com/Shopify/polaris/assets/18447883/e11b06fc-af30-43fc-8316-f7455e022a46) | ![Screenshot 2023-12-13 at 12 30 12 PM](https://github.com/Shopify/polaris/assets/18447883/e11b06fc-af30-43fc-8316-f7455e022a46)|
|hover| ![Screenshot 2023-12-13 at 12 24 36 PM](https://github.com/Shopify/polaris/assets/18447883/bfe347be-0743-49f2-9d2f-3c3a72a9ec46) | ![Screenshot 2023-12-13 at 12 25 16 PM](https://github.com/Shopify/polaris/assets/18447883/525544da-dc99-49e2-acb2-6a6ba114efcf) |
|pressed/active| ![Screenshot 2023-12-13 at 12 20 16 PM](https://github.com/Shopify/polaris/assets/18447883/21b81da8-f89d-45f8-90a2-8c25bbd63494) | ![Screenshot 2023-12-13 at 12 21 06 PM](https://github.com/Shopify/polaris/assets/18447883/093ae019-517b-4a5d-8507-e19fe2b10595)|

heyjoethomas replied:
Yes. This makes more sense.

chloerice commented on polaris-react/src/components/Button/Button.scss:
```diff
@@ -1,1043 +1,352 @@
-      line-height: var(--p-font-line-height-400);
-    }
-  }
+.iconOnly:is(.variantTertiary) {
```
(Icon only plain buttons are secondary too) 

```suggestion
.iconOnly:is(.variantTertiary, .variantPlain) {
```

| Before | After |
|--------|--------|
| ![Screenshot 2023-12-13 at 12 54 34 PM](https://github.com/Shopify/polaris/assets/18447883/b2088ba9-4b4c-4ad4-adc0-6dea53ecf1e4) | ![Screenshot 2023-12-13 at 12 53 42 PM](https://github.com/Shopify/polaris/assets/18447883/b3dd5e58-1d37-4dda-a98d-0633572967da) |

sam-b-rose replied:
I did notice this, but it seemed confusing to me why the `plain` icon button wouldn't be blue. Would it make more sense to use the `monochromePlain` or `tertiary`? 🤔 

@heyjoethomas should we continue this pattern or could we consider this a "fix" by retaining the blue link color for iconOnly buttons?

sam-b-rose replied:
Confirmed with Joe, the less risky route is to keep the color secondary as it is in `main` currently.

I still think this design choice is confusing and we should perhaps revisit the decision in a separate issue 👍

chloerice commented on polaris-react/src/components/Button/Button.scss:
```diff
@@ -1,1043 +1,352 @@
-
-  &.iconOnly {
-    margin: 0;
-    padding: 0;
```
The `.iconOnly:is(.variantPlain, .variantMonochromePlain)` `margin: 0;` and `padding: 0;` needs to be ported over, without it there's an alignment regression in the plain variant.

| Before | After |
|--------|--------|
|![Screenshot 2023-12-13 at 3 29 59 PM](https://github.com/Shopify/polaris/assets/18447883/c6520e32-b323-48b0-a34b-e82dc19b347a) |![Screenshot 2023-12-13 at 3 34 44 PM](https://github.com/Shopify/polaris/assets/18447883/de04252e-acc3-4156-8153-67dad5248586)|

sam-b-rose committed:
Update iconOnly color and spacing

sam-b-rose committed:
Add state colors to plain and tone

chloerice reviewed (approved):
:shipit:

aveline reviewed (approved):
🎉 Nice work!

sam-b-rose committed:
Apply negative margin for tertiary icon-only buttons

sam-b-rose committed:
'Lighten shadow for active/pressed critical and success tones

sam-b-rose committed:
Keep normal padding for icons with text and disclosure

sam-b-rose committed:
Use larger font size for plain and monochrome

sam-b-rose committed:
Use min-height instead of height
